src/library/scala/collection/mutable/HashSet.scala

```diff
@@ -28,64 +29,233 @@ package mutable
   * @define mayNotTerminateInf
   * @define willNotTerminateInf
   */
-final class HashSet[A]
+final class HashSet[A](initialCapacity: Int, loadFactor: Double)
   extends AbstractSet[A]
     with SetOps[A, HashSet, HashSet[A]]
     with StrictOptimizedIterableOps[A, HashSet, HashSet[A]] {
 
-  @transient private[this] var table = new FlatHashTable[A]
+  def this() = this(HashSet.defaultInitialCapacity, HashSet.defaultLoadFactor)
 
-  // Used by scala-java8-compat (private[mutable] erases to public, so Java code can access it)
-  private[mutable] def getTable: FlatHashTable[A] = table
-  override def iterator: Iterator[A] = table.iterator
+  import HashSet.Node
 
-  override def iterableFactory: IterableFactory[HashSet] = HashSet
+  /** The actual hash table. */
+  private[this] var table = new Array[Node[A]](tableSizeFor(initialCapacity))
+
+  /** The next size value at which to resize (capacity * load factor). */
+  private[this] var threshold: Int = newThreshold(table.length)
+
+  private[this] var contentSize = 0
+
+  override def size: Int = contentSize
 
-  def addOne(elem: A): this.type = {
-    table.addElem(elem)
-    this
+  @`inline` private[this] def computeHash(o: A): Int = {
+    val h = o.##
+    // Improve the hash by xoring the high 16 bits into the low 16 bits just in case entropy is skewed towards the
+    // high-value bits. We only use the lowest bits to determine the hash bucket. This is the same improvement
+    // algorithm as in java.util.HashMap.
+    h ^ (h >>> 16)
   }
-  def subtractOne(elem: A): this.type = {
-    table.removeElem(elem)
-    this
+
+  @`inline` private[this] def index(hash: Int) = hash & (table.length - 1)
+
+  override def contains(elem: A): Boolean = findNode(elem) ne null
+
+  @`inline` private[this] def findNode(elem: A): Node[A] = {
+    val hash = computeHash(elem)
+    table(index(hash)) match {
+      case null => null
+      case nd => nd.findNode(elem, hash)
+    }
   }
 
-  override def add(elem: A): Boolean = table.addElem(elem)
+  override def sizeHint(size: Int): Unit = {
+    val target = tableSizeFor(((size + 1).toDouble / loadFactor).toInt)
+    if(target > table.length) growTable(target)
+  }
 
-  override def remove(elem: A): Boolean = table.removeElem(elem)
+  override def add(elem: A) : Boolean = {
+    if(contentSize + 1 >= threshold) growTable(table.length * 2)
+    addElem(elem, computeHash(elem))
+  }
 
-  def clear(): Unit = table.clearTable()
+  override def addAll(xs: IterableOnce[A]): this.type = {
+    sizeHint(xs.knownSize)
+    super.addAll(xs)
+  }
```

Comment:
We could take advantage of the cached hashcodes in both this implementation of `mutable.HashSet` as well as in `immutable.HashSet`.

Maybe an other optimization:
 
When `case xs: mutable.HashMap[A] if xs.table.length >= this.table.length`, then each of those buckets maps onto exactly one bucket in `this.table`. Maybe each bucket in that could be "merge joined" into the bucket in `this`, since they're both sorted linked lists? Maybe it would get complicated with the extra housekeeping of making sure to resize the table when `this` gets too big

Reply:
You could check for equal table sizes after growing this table. If you check before growing and delay growing until you've added the new data you could end up adding a lot of data in an inefficient way to linked lists even though you know you'll have to split them afterwards.

Reply:
Hmm but whether or not this table should grow depends on the overlap of the two HashMaps. It's probably not worth doing I guess.